File: Python/mditre/core/base_layer.py
"""
Base layer class for MDITRE architecture
Provides common interface for all layers to enable modularity and extensibility
"""

import torch
import torch.nn as nn
from abc import ABC, abstractmethod
from typing import Dict, Any, Optional, Tuple, Union


class BaseLayer(nn.Module, ABC):
    """
    Abstract base class for all MDITRE layers
    
    This provides a common interface that all layers must implement,
    enabling:
    - Easy layer swapping and experimentation
    - Consistent parameter initialization
    - Standardized forward pass interface
    - Future layer additions without breaking existing code
    """
    
    def __init__(self, layer_name: str, layer_config: Dict[str, Any]):
        """
        Initialize base layer
        
        Args:
            layer_name: Human-readable name for this layer
            layer_config: Configuration dictionary for layer parameters
        """
        super(BaseLayer, self).__init__()
        self.layer_name = layer_name
        self.layer_config = layer_config
        self._setup_complete = False
        
    @abstractmethod
    def forward(self, *args, **kwargs) -> Union[torch.Tensor, Tuple[torch.Tensor, ...]]:
        """
        Forward pass through the layer
        
        Args:
            *args: Positional arguments (typically input tensors)
            **kwargs: Additional layer-specific keyword arguments
            
        Returns:
            Output tensor(s) - can be a single Tensor or Tuple of Tensors
        """
        pass
    
    @abstractmethod
    def init_params(self, init_args: Dict[str, Any]) -> None:
        """
        Initialize layer parameters
        
        Args:
            init_args: Dictionary containing initialization values
        """
        pass
    
    def get_config(self) -> Dict[str, Any]:
        """
        Get layer configuration
        
        Returns:
            Configuration dictionary
        """
        return self.layer_config.copy()
    
    def get_layer_info(self) -> Dict[str, Any]:
        """
        Get detailed information about this layer
        
        Returns:
            Dictionary with layer metadata
        """
        return {
            'name': self.layer_name,
            'type': self.__class__.__name__,
            'config': self.get_config(),
            'num_parameters': sum(p.numel() for p in self.parameters()),
            'trainable_parameters': sum(p.numel() for p in self.parameters() if p.requires_grad)
        }
    
    def __repr__(self) -> str:
        """String representation of the layer"""
        return f"{self.__class__.__name__}(name='{self.layer_name}')"


class LayerRegistry:
    """
    Registry for dynamically managing available layers
    
    This allows for:
    - Runtime layer selection
    - Easy addition of new layer implementations
    - Version control of layer implementations
    """
    
    _registry = {
        'layer1': {},  # Phylogenetic Focus layers
        'layer2': {},  # Temporal Focus layers  
        'layer3': {},  # Detector layers
        'layer4': {},  # Rule layers
        'layer5': {},  # Classification layers
    }
    
    @classmethod
    def register(cls, layer_type: str, layer_name: str):
        """
        Decorator to register a layer implementation
        
        Args:
            layer_type: Type of layer (layer1, layer2, etc.)
            layer_name: Unique name for this implementation
            
        Example:
            @LayerRegistry.register('layer1', 'spatial_agg_dynamic')
            class SpatialAggDynamic(BaseLayer):
                ...
        """
        def decorator(layer_class):
            cls._registry[layer_type][layer_name] = layer_class
            return layer_class
        return decorator
    
    @classmethod
    def get_layer(cls, layer_type: str, layer_name: str):
        """
        Get a registered layer class
        
        Args:
            layer_type: Type of layer
            layer_name: Name of specific implementation
            
        Returns:
            Layer class
        """
        if layer_type not in cls._registry:
            raise ValueError(f"Unknown layer type: {layer_type}")
        if layer_name not in cls._registry[layer_type]:
            raise ValueError(f"Unknown layer name '{layer_name}' for type '{layer_type}'")
        return cls._registry[layer_type][layer_name]
    
    @classmethod
    def list_layers(cls, layer_type: Optional[str] = None) -> Dict[str, list]:
        """
        List all registered layers
        
        Args:
            layer_type: Specific layer type to list, or None for all
            
        Returns:
            Dictionary of available layers
        """
        if layer_type is not None:
            return {layer_type: list(cls._registry[layer_type].keys())}
        return {k: list(v.keys()) for k, v in cls._registry.items()}

File: R/R/base_layer.R
#' Base Layer for MDITRE Architecture
#'
#' @description
#' Abstract base class for all MDITRE layers. Provides common interface
#' enabling modularity and extensibility. All MDITRE layers inherit from this.
#'
#' @section Methods:
#' \describe{
#'   \item{initialize(layer_name, layer_config)}{Initialize base layer}
#'   \item{forward(...)}{Forward pass through the layer (must be implemented by subclasses)}
#'   \item{init_params(init_args)}{Initialize layer parameters}
#'   \item{get_config()}{Get layer configuration}
#'   \item{get_layer_info()}{Get detailed information about this layer}
#' }
#'
#' @param layer_name Character string with human-readable name for this layer
#' @param layer_config Named list with configuration parameters for the layer
#'
#' @export
#' @examples
#' \dontrun{
#' # Custom layer inheriting from base_layer
#' my_layer <- nn_module(
#'   "MyLayer",
#'   inherit = base_layer,
#'   
#'   initialize = function(layer_name, layer_config) {
#'     super$initialize(layer_name, layer_config)
#'     # Add custom initialization here
#'   },
#'   
#'   forward = function(x) {
#'     # Implement forward pass
#'     return(x)
#'   }
#' )
#' }
base_layer <- nn_module(
  "BaseLayer",
  
  #' @description Initialize base layer
  #' @param layer_name Human-readable name for this layer
  #' @param layer_config Configuration list for layer parameters
  initialize = function(layer_name, layer_config) {
    self$layer_name <- layer_name
    self$layer_config <- layer_config
    self$setup_complete <- FALSE
  },
  
  #' @description Forward pass through the layer (must be overridden)
  #' @param ... Arguments passed to forward method
  forward = function(...) {
    stop(sprintf("Subclass %s must implement forward() method", class(self)[1]))
  },
  
  #' @description Initialize layer parameters (must be overridden)
  #' @param init_args Named list containing initialization values
  init_params = function(init_args) {
    stop(sprintf("Subclass %s must implement init_params() method", class(self)[1]))
  },
  
  #' @description Get layer configuration
  #' @return Named list with configuration
  get_config = function() {
    return(self$layer_config)
  },
  
  #' @description Get detailed information about this layer
  #' @return Named list with layer metadata including name, type, config, and parameter counts
  get_layer_info = function() {
    # Count parameters
    total_params <- 0
    trainable_params <- 0
    
    for (param_name in names(self$parameters)) {
      param <- self$parameters[[param_name]]
      if (inherits(param, "torch_tensor")) {
        param_count <- param$numel()$item()
        total_params <- total_params + param_count
        if (param$requires_grad) {
          trainable_params <- trainable_params + param_count
        }
      }
    }
    
    return(list(
      name = self$layer_name,
      type = class(self)[1],
      config = self$get_config(),
      num_parameters = total_params,
      trainable_parameters = trainable_params
    ))
  }
)


#' Layer Registry for MDITRE
#'
#' @description
#' Registry for dynamically managing available layers. Allows for runtime
#' layer selection, easy addition of new layer implementations, and version
#' control of layer implementations.
#'
#' @export
LayerRegistry <- R6::R6Class(
  "LayerRegistry",
  
  public = list(
    #' @field registry Named list storing registered layer classes
    registry = list(),
    
    #' @description Register a new layer type
    #' @param name Character string name for the layer
    #' @param layer_class The nn_module class definition
    #' @param version Character string version (default: "1.0.0")
    register = function(name, layer_class, version = "1.0.0") {
      if (name %in% names(self$registry)) {
        warning(sprintf("Overwriting existing layer '%s'", name))
      }
      
      self$registry[[name]] <- list(
        class = layer_class,
        version = version,
        registered_at = Sys.time()
      )
      
      message(sprintf("Registered layer: %s (v%s)", name, version))
    },
    
    #' @description Get a registered layer class
    #' @param name Character string name of the layer
    #' @return The layer class definition
    get = function(name) {
      if (!name %in% names(self$registry)) {
        stop(sprintf("Layer '%s' not found in registry. Available layers: %s",
                     name, paste(names(self$registry), collapse = ", ")))
      }
      return(self$registry[[name]]$class)
    },
    
    #' @description List all registered layers
    #' @return Character vector of registered layer names
    list_layers = function() {
      return(names(self$registry))
    },
    
    #' @description Get information about a registered layer
    #' @param name Character string name of the layer
    #' @return Named list with layer metadata
    get_info = function(name) {
      if (!name %in% names(self$registry)) {
        stop(sprintf("Layer '%s' not found in registry", name))
      }
      
      layer_entry <- self$registry[[name]]
      return(list(
        name = name,
        version = layer_entry$version,
        registered_at = layer_entry$registered_at
      ))
    },
    
    #' @description Remove a layer from registry
    #' @param name Character string name of the layer
    remove = function(name) {
      if (name %in% names(self$registry)) {
        self$registry[[name]] <- NULL
        message(sprintf("Removed layer: %s", name))
      } else {
        warning(sprintf("Layer '%s' not found in registry", name))
      }
    }
  )
)

# Create global layer registry instance
#' @export
layer_registry <- LayerRegistry$new()
